tests/postgres_tests/test_serial.py

```diff
@@ -0,0 +1,344 @@
+        self.assertEqual(serial.cast_db_type(connection), "integer")
+        self.assertEqual(big_serial.cast_db_type(connection), "bigint")
+
+    def test_rel_db_types(self):
+        class SerialPK(models.Model):
+            id = SerialField(primary_key=True)
+
+        class SmallSerialPK(models.Model):
+            id = SmallSerialField(primary_key=True)
+
+        class BigSerialPK(models.Model):
+            id = BigSerialField(primary_key=True)
+
+        class SerialFK(models.Model):
+            fk = models.ForeignKey(SerialPK, on_delete=models.CASCADE)
+
+        class SmallSerialFK(models.Model):
+            fk = models.ForeignKey(SmallSerialPK, on_delete=models.CASCADE)
+
+        class BigSerialFK(models.Model):
+            fk = models.ForeignKey(BigSerialPK, on_delete=models.CASCADE)
+
+        fk = SerialFK._meta.get_field("fk")
+        small_fk = SmallSerialFK._meta.get_field("fk")
+        big_fk = BigSerialFK._meta.get_field("fk")
+
+        self.assertEqual(fk.rel_db_type(connection), "integer")
+        self.assertEqual(small_fk.rel_db_type(connection), "smallint")
+        self.assertEqual(big_fk.rel_db_type(connection), "bigint")
+
+
+class SerialModelBaseTests(PostgreSQLTestCase):
+    @classmethod
+    def reset_sequences_to_1(cls, model):
+        with connection.cursor() as cursor:
+            for statement in cls.get_sequence_reset_by_name_sql(model):
+                cursor.execute(statement)
+
+    @classmethod
+    def get_sequence_reset_by_name_sql(cls, model):
+        return connection.ops.sequence_reset_by_name_sql(
+            no_style(), cls.get_sequences(model)
+        )
+
+    @classmethod
+    def get_sequences(cls, model):
+        with connection.cursor() as cursor:
+            return connection.introspection.get_sequences(cursor, model._meta.db_table)
+
+    @classmethod
+    def reset_sequences_to_max(cls, model):
+        with connection.cursor() as cursor:
+            for statement in cls.get_sequence_reset_sql(model):
+                cursor.execute(statement)
+
+    @classmethod
+    def get_sequence_reset_sql(cls, model):
+        return connection.ops.sequence_reset_sql(no_style(), [model])
+
+    @classmethod
+    def get_table_description(cls, model):
+        with connection.cursor() as cursor:
+            return connection.introspection.get_table_description(
+                cursor, model._meta.db_table
+            )
+
+
+class SerialModelTests(SerialModelBaseTests):
+    def test_create(self):
+        self.reset_sequences_to_1(SerialModel)
+
+        obj_1 = SerialModel.objects.create()
+        self.assertEqual(obj_1.small_serial, 1)
+        self.assertEqual(obj_1.serial, 1)
+        self.assertEqual(obj_1.big_serial, 1)
+
+        obj_2 = SerialModel.objects.create()
+        self.assertEqual(obj_2.small_serial, 2)
+        self.assertEqual(obj_2.serial, 2)
+        self.assertEqual(obj_2.big_serial, 2)
+
+        obj_3 = SerialModel.objects.create(
```

Comment:
I think we should either break this test up into smaller cases with descriptive test names for what we're checking or add one or two comments. I was trying to see if any of the object creation is unnessary

This is maybe a suggestion, we can probably do better
```python
        # When setting to None, will continue the sequence.
        obj_3 = SerialModel.objects.create(
            small_serial=None,
            serial=None,
            big_serial=None,
        )
        self.assertEqual(obj_3.small_serial, 3)
        self.assertEqual(obj_3.serial, 3)
        self.assertEqual(obj_3.big_serial, 3)

        # Serial fields do not need to be unique.
        manual_obj_1 = SerialModel.objects.create(
            small_serial=1,
            serial=1,
            big_serial=1,
        )
        self.assertEqual(manual_obj_1.small_serial, 1)
        self.assertEqual(manual_obj_1.serial, 1)
        self.assertEqual(manual_obj_1.big_serial, 1)

        # The sequence is not interrupted by other objects.
        manual_obj_5 = SerialModel.objects.create(
            small_serial=5,
            serial=5,
            big_serial=5,
        )
        self.assertEqual(manual_obj_5.small_serial, 5)
        self.assertEqual(manual_obj_5.serial, 5)
        self.assertEqual(manual_obj_5.big_serial, 5)

        obj_4 = SerialModel.objects.create()
        self.assertEqual(obj_4.small_serial, 4)
        self.assertEqual(obj_4.serial, 4)
        self.assertEqual(obj_4.big_serial, 4)

        obj_5 = SerialModel.objects.create()
        self.assertEqual(obj_5.small_serial, 5)
        self.assertEqual(obj_5.serial, 5)
        self.assertEqual(obj_5.big_serial, 5)
```

Reply:
I added some comments, let me know if that's acceptable?

Reply:
Yes, I like it thank you 👍